core/src/main/java/org/elasticsearch/cluster/MasterNodeChangePredicate.java

```diff
@@ -33,13 +33,13 @@ private MasterNodeChangePredicate() {
      */
     public static Predicate<ClusterState> build(ClusterState currentState) {
         final long currentVersion = currentState.version();
-        final String currentMaster = currentState.nodes().getMasterNodeId();
+        final String currentMasterId = currentState.nodes().getMasterNode().getEphemeralId();
         return newState -> {
-            final String newMaster = newState.nodes().getMasterNodeId();
+            final String newMasterId = newState.nodes().getMasterNode().getEphemeralId();
             final boolean accept;
-            if (newMaster == null) {
+            if (newMasterId == null) {
```

Comment:
`newMasterId` (= `getEphemeralId()`) cannot be null, `getMasterNode()` can be, however.